torch_xla/experimental/scan.py

```diff
@@ -82,14 +88,335 @@ def scan(fn, init, xs):
   if xs_length is None:
     raise ValueError(f"`xs` {xs} is an empty PyTree.")
 
-  carry = init
-  ys = []
+  carry, carry_history, ys = Scan.apply(fn, init, xs)  # type: ignore
+  return carry, ys
+
 
-  for i in range(xs_length):
-    carry, y = fn(carry, tree_map(lambda x: x[i], xs))
-    ys.append(y)
+def dynamic_update_slice(ys: xb.Op, y: xb.Op, idx: xb.Op) -> xb.Op:
+  # See https://openxla.org/xla/operation_semantics#dynamicupdateslice.
+  y = y.broadcast([1])
+  indices = [idx]
+  for _ in range(ys.shape().rank - 1):
+    indices.append(idx.zeros_like())
+  return ys.dynamic_update_slice(y, indices)
 
-  # Combine the list of PyTrees into one PyTree, where the leaves are
-  # stacked into a new major axis.
-  ys = tree_map(lambda *x: torch.stack(x), *ys)
-  return carry, ys
+
+def dynamic_slice(xs: xb.Op, idx: xb.Op) -> xb.Op:
+  indices = [idx]
+  for _ in range(xs.shape().rank - 1):
+    indices.append(idx.zeros_like())
+  slice_shape = list(xs.shape().sizes)
+  slice_shape[0] = 1
+  sliced = xs.dynamic_slice(indices, slice_shape)
+  shape = list(xs.shape().sizes)
+  shape = shape[1:]
+  return sliced.reshape(shape)
+
+
+class Builder:
+
+  def __init__(self, name: str):
+    self._builder = xb.create_builder(name)
+    self._params = []
+    self._param_tensors = []
+
+  def add_param(self, val: torch.Tensor):
+    idx = len(self._params)
+    param = xb.mkparam(self._builder, idx, xb.tensor_shape(val))
+    self._params.append(param)
+    self._param_tensors.append(val)
+    return idx
+
+  def params(self) -> Tuple[xb.Op, ...]:
+    return tuple(self._params)
+
+  def param_tensors(self) -> Tuple[torch.Tensor, ...]:
+    return tuple(self._param_tensors)
+
+  def num_params(self) -> int:
+    return len(self._params)
+
+
+def _scan_impl(fn, init, xs):
+  """Forward logic of scan without gradient tracking.
+  
+  See the `Scan` class which implements an autograd `Function` and builds
+  autograd support on top of `_scan_impl`.
+  """
+
+  flat_init, carry_spec = tree_flatten(init)
+  flat_xs, xs_spec = tree_flatten(xs)
+
+  # Because `flat_fn` returns a concatenated flattened carry and y list,
+  # we need to know how many elements out of that list is the carry.
+  flat_carry_len = len(flat_init)
+  flat_xs_len = len(flat_xs)
+
+  # `fn` operates on PyTrees and returns PyTrees. However, XLA only understands
+  # (lists or tuples of) tensors. So we will craft a `flat_fn` that takes in
+  # flattened PyTrees, internally recreates the desired tree structure, then calls `fn`.
+  def flat_fn(*seq: torch.Tensor) -> Tuple[PyTree, PyTree]:
+    carry = seq[:flat_carry_len]
+    x = seq[flat_carry_len:]
+    carry_pytree = tree_unflatten(carry, carry_spec)
+    x_pytree = tree_unflatten(x, xs_spec)
+    return fn(carry_pytree, x_pytree)
+
+  # Abstractly trace and lower `fn`.
```

Comment:
Note to self: To further ensure the layers perform identically structured computations, we can trace each layer and compare their HLO. Tracing should be faster than end-to-end compilation. In practice this means scan would trace i times given input sequence of length i.

We can still avoid re-tracing the backward operations multiple times. We assume that identical forward graphs will produce identical backward graphs.

Reply:
Chatted offline with Jack -- we'll just note this limitation in comments.